lib/puppet/pops/types/type_calculator.rb

```diff
@@ -313,240 +225,111 @@ def equals(left, right)
   # @api public
   #
   def type(c)
-    raise ArgumentError, "Argument must be a Class" unless c.is_a? Class
+    raise ArgumentError, 'Argument must be a Class' unless c.is_a? Class
 
     # Can't use a visitor here since we don't have an instance of the class
     case
     when c <= Integer
-      type = Types::PIntegerType.new()
+      type = Types::PIntegerType::DEFAULT
     when c == Float
-      type = Types::PFloatType.new()
+      type = Types::PFloatType::DEFAULT
     when c == Numeric
-      type = Types::PNumericType.new()
+      type = Types::PNumericType::DEFAULT
     when c == String
-      type = Types::PStringType.new()
+      type = Types::PStringType::DEFAULT
     when c == Regexp
-      type = Types::PRegexpType.new()
+      type = Types::PRegexpType::DEFAULT
     when c == NilClass
-      type = Types::PUndefType.new()
+      type = Types::PUndefType::DEFAULT
     when c == FalseClass, c == TrueClass
-      type = Types::PBooleanType.new()
+      type = Types::PBooleanType::DEFAULT
     when c == Class
-      type = Types::PType.new()
+      type = Types::PType::DEFAULT
     when c == Array
       # Assume array of data values
-      type = Types::PArrayType.new()
-      type.element_type = Types::PDataType.new()
+      type = Types::PArrayType::DATA
     when c == Hash
       # Assume hash with scalar keys and data values
-      type = Types::PHashType.new()
-      type.key_type = Types::PScalarType.new()
-      type.element_type = Types::PDataType.new()
-    else
-      type = Types::PRuntimeType.new(:runtime => :ruby, :runtime_type_name => c.name)
+      type = Types::PHashType::DATA
+   else
+      type = Types::PRuntimeType.new(:ruby, c.name)
     end
     type
   end
 
-  # Generalizes value specific types. The given type is mutated and returned.
+  # Generalizes value specific types. The generalized type is returned.
   # @api public
-  def generalize!(o)
-    @@generalize_visitor.visit_this_0(self, o)
-    o.eAllContents.each { |x| @@generalize_visitor.visit_this_0(self, x) }
+  def generalize(o)
+    o = o.generalize if o.is_a?(Types::PAnyType)
     o
```

Comment:
Slightly more efficient to do: `o.is_a?(Type::PAnyType) ? o.generalize : o` as it avoids the assignment to o